lib/spack/spack/build_environment.py

```diff
@@ -321,12 +321,32 @@ def set_build_environment_variables(pkg, env, dirty=False):
         extra_rpaths = ':'.join(compiler.extra_rpaths)
         env.set('SPACK_COMPILER_EXTRA_RPATHS', extra_rpaths)
 
-    # Add bin directories from dependencies to the PATH for the build.
-    bin_dirs = reversed(filter(os.path.isdir, [
-        '%s/bin' % d.prefix for d in pkg.spec.dependencies(deptype='build')]))
+    # Function to add a directory to a list only if
+    # that directory exists and is not already in the list
+    def add_dir_to_list(bin_dir, list):
+        if os.path.isdir(bin_dir):
+            if bin_dir not in list:
+                list.append(bin_dir)
+
+    # Create a list which will contain the bin directories
+    bin_dirs = []
+
+    # Add all bin directories which build dependencies and their
+    # run dependencies require.
+    for bdep in [ d for d in pkg.spec.dependencies(deptype='build') ]:
```

Comment:
Isn't this line equivalent to the simpler

`for bdep in pkg.spec.dependencies(deptype='build'):`